### PR TITLE
fixed 2 known OOP issues

### PR DESCRIPTION
**Pull request by heejaechang**
1. there is a known issue where RequestServiceAsync might fail randomly.
   people are looking at the root cause, but until that is fixed,
   we will retry rather than failing right away.

2. if we close named pipe stream in the middle of object writer/reader is
   using them, we can get various exceptions. make sure we check cancellation
   before propagating that exception.

### CODE REVIEW

**Comment by heejaechang** [comment]
retest this please

**Comment by heejaechang** [comment]
@srivatsn @mavasani @panopticoncentral can you take a look?

**Comment by heejaechang** [comment]
retest  windows_debug_unit32_prtest  please

File: src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.cs [hunk 1]
```diff
@@ -105,16 +106,33 @@ private void OnRpcDisconnected(object sender, JsonRpcDisconnectedEventArgs e)
...
+            // call to get service can fail due to this bug - devdiv#288961
+            // until root cause is fixed, we decide to have retry rather than fail right away
+            for (var i = 0; i < max_retry; i++)
             {
-                return await client.RequestServiceAsync(serviceName, cancellationToken).ConfigureAwait(false);
+                cancellationToken.ThrowIfCancellationRequested();
+
+                try
+                {
+                    var descriptor = new ServiceDescriptor(serviceName) { HostGroup = new HostGroup(hostGroup) };
```
Comment (heejaechang) [inline src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.cs]: @AArnott @IlyaBiryukov is this retry looks fine?

Comment (IlyaBiryukov) [inline src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.cs]: Looks fine

Comment (IlyaBiryukov) [inline src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.cs]: Looks fine

Comment (IlyaBiryukov) [inline src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.cs]: Yep, looks fine.

**Comment by heejaechang** [comment]
ping?

File: src/Workspaces/Remote/ServiceHub/Shared/Extensions.cs [hunk 2]
```diff
@@ -138,25 +138,27 @@ internal static partial class Extensions
...
-        private static bool IsCancelled(Exception ex)
+        private static bool IsCancelled(Exception ex, CancellationToken token)
         {
-            // object disposed exception can be thrown from StreamJsonRpc if JsonRpc is disposed in the middle of read/write.
-            // the way we added cancellation support to the JsonRpc which doesn't support cancellation natively
-            // can cause this exception to happen. newer version supports cancellation token natively, but
-            // we can't use it now, so we will catch object disposed exception and check cancellation token
-            if (ex is ObjectDisposedException || ex is OperationCanceledException)
+            // check whether we are in cancellation mode
+            if (token.IsCancellationRequested)
```
Comment (mavasani) [inline src/Workspaces/Remote/ServiceHub/Shared/Extensions.cs]: This seems a bit weird. You are checking for cancellation only using the token, but the general mechanism is to check using the exception type. Basically we are swallowing all exceptions if cancellation happened?

File: src/Workspaces/Remote/ServiceHub/Shared/Extensions.cs [hunk 3]
```diff
@@ -138,25 +138,27 @@ internal static partial class Extensions
 
                     return result;
                 }
-            }
-            catch (Exception ex) when (IsCancelled(ex))
-            {
-                cancellationToken.ThrowIfCancellationRequested();
-                throw;
+                catch (Exception ex) when (IsCancelled(ex, mergedCancellation.Token))
```
Comment (heejaechang) [inline src/Workspaces/Remote/ServiceHub/Shared/Extensions.cs]: ya, it is linked cancellationtoken. so merged one will be raised when cancellation token is raised.

File: src/Workspaces/Remote/ServiceHub/Shared/Extensions.cs (hunk 2 above)
Comment (heejaechang) [inline src/Workspaces/Remote/ServiceHub/Shared/Extensions.cs]: yep. that is the purpose of this change. when cancellation is raised, various exception can happen based on what was happening when we forcefully closed named pipe stream.

File: src/Workspaces/Remote/ServiceHub/Shared/Extensions.cs [hunk 4]
```diff
@@ -138,25 +138,27 @@ internal static partial class Extensions
...
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+                    throw;
+                }
             }
         }
 
-        private static bool IsCancelled(Exception ex)
+        private static bool IsCancelled(Exception ex, CancellationToken token)
```
Comment (AArnott) [inline src/Workspaces/Remote/ServiceHub/Shared/Extensions.cs]: This method seems poorly named. `IsCanceled` sounds very benign (and the exception type seems unrelated unless it were somehow included in the cancellation check). The fact that the exception crashes the product if it isn't canceled would be totally unexpected given the method name and lack of xml doc comments to help call out otherwise.
Perhaps simply "ThrowIfNotCanceled"?

**Comment by AArnott** [comment]
> If we close named pipe stream in the middle of object writer/reader is
using them, we can get various exceptions. make sure we check cancellation
before propagating that exception.

This is interesting. Who is throwing these exceptions? Is it newtonsoft.json? If so, or anything my team owns, I'd be interested in seeing some sampling of callstacks and/or exception types if you have them as this might be something we want to fix at a lower level as well.

**Comment by heejaechang** [comment]
@AArnott named pipe here has nothing to do with service hub. it is our own named pipe which we create so that we don't use expensive json and use binary directly.
